query-planner-js/src/utilities/debug.ts

```diff
@@ -0,0 +1,143 @@
+// Simple debugging facility.
+
+import chalk from 'chalk';
+
+function indentString(indentLevel: number) : string {
+  let str = "";
+  for (let i = 0; i < indentLevel; i++) {
+    str += chalk.blackBright("⎸ ");
+  }
+  return str;
+}
+
+/**
+ * A facility to log messages for debugging.
+ *
+ * A `DebugLogger` maintains an indentation level that is used before logged messages. That indentation level gets
+ * increased by calling `group()` and decreased by `groupEnd()`. This is meant to help representing the nestedness
+ * of the code being debugged.
+ */
+export class DebugLogger {
+  private currentIndentLevel: number = 0;
+
+  /**
+   * Builds a new `DebugLogger`.
+   *
+   * @param enabled - whether the logger is enabled. If it is disabled, no message will be logged. Currently, a logger
+   * cannot be enabled/disabled after construction.
+   */
+  constructor(private readonly enabled: boolean) {}
+
+  /**
+   * Logs the provided message at the current indentation level, if this logger is enabled.
+   *
+   * @param message - the message to log. This can optionally be a closure, which allows for the message to only be
+   * computed if the logger is enabled (the closure is only called in this case).
+   * @param prefix - an optional string printed as a direct prefix of the message.
+   */
+  public log(message: string | (() => string), prefix: string = chalk.yellow('• ')) {
+    if (!this.enabled) return this;
+
+    if (typeof message !== 'string') {
+      message = message();
+    }
+    const indentStr = indentString(this.currentIndentLevel);
+    console.log(indentStr + prefix + message);
+    return this;
+  }
+
+  /**
+   * Logs the provided values, indented and one per line.
+   *
+   * @param values - an array of the values to log.
+   * @param fct - the function to apply to each of the value to convert it to a string to log.
+   */
+  public groupedValues<T>(values: T[], fct: (v: T) => string) {
+    if (!this.enabled) return this;
+
+    this.group();
+    const indentStr = indentString(this.currentIndentLevel);
+    for (const value of values) {
+      console.log(indentStr + '- ' + fct(value));
+    }
+    return this.groupEnd();
+  }
+
+  /**
+   * Logs the entries of the provided map, indented and one per line.
+   *
+   * @param map - a map of keys and values to log.
+   * @param keyFct - the function to apply to each of the key to convert it to a string to log.
+   * @param valueFct - the function to apply to each of the value to convert it to a string to log.
```

Comment:
I think `fn` would be the more common abbreviation for function. So maybe `keyFn` and `valueFn`? Or even `keyPrintFn` and `valuePrintFn`?